test/test_main.py

```diff
@@ -0,0 +1,178 @@
+#!/usr/bin/env python3
+
+import unittest
+import logging
+import os
+import sys
+import subprocess
+
+from unittest.mock import (
+    patch,
+)
+
+JUJU_CHARM_DIR = os.path.join(
+    os.path.dirname(os.path.abspath(__file__)),
+    'charms/main'
+)
+
+# change cwd for the current process to the test charm directory
+os.chdir(JUJU_CHARM_DIR)
+os.environ['JUJU_CHARM_DIR'] = JUJU_CHARM_DIR
+
+del sys.path[0]
+sys.path.insert(0, os.path.join(JUJU_CHARM_DIR, 'lib'))
```

Comment:
All of the logic above cannot be here. This file is a usual Python module like any other, so we cannot have actions that would introduce side effects by simply importing the module.